### PR TITLE
Permits cartesians to use delta-grid levelling strategy

### PR DESCRIPTION
**Pull request by lordofhyphens**
adds 3 new config options to leveling-strategy.delta-grid:
max_x - defaults to value of grid_radius - Maximum travel from center of bed in +/-X direction 
max_y - defaults to value of grid_radius - Maximum travel from center of bed in +/-Y direction.
do_home - default True - Home at beginning of G31 call.
is_square - default False - turns off the check to not probe the corners (do not set on a delta with a round bed)

Caveat: 0,0 must be the center of the bed when using this leveling strategy.

### CODE REVIEW

File: src/modules/tools/zprobe/DeltaGridStrategy.h [hunk 1]
```diff
@@ -40,8 +40,15 @@ class DeltaGridStrategy : public LevelingStrategy
...
     std::tuple<float, float, float> probe_offsets;
     uint8_t grid_size;
+    float x_max,y_max;
 
     struct {
         bool save:1;
     };
+    struct {
+        bool do_home:1;
+    };
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.h]: this goes in the one struct.
```
struct {
          bool save:1;
          bool do_home:1;
 };
```

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 2]
```diff
@@ -109,6 +113,17 @@ bool DeltaGridStrategy::handleConfig()
...
+    y_max = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, y_max_checksum)->by_default(0.0F)->as_number();
+    do_home = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, do_home_checksum)->by_default(true)->as_bool();
+    is_square = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, is_square_checksum)->by_default(false)->as_bool();
+
+    // intelligently set defaults.
+    if (x_max >= 1.0F) grid_radius = x_max;
+    if (x_max < 1.0F) x_max = grid_radius;
+    if (y_max >= 1.0F) grid_radius = y_max;
+    if (y_max < 1.0F) y_max = grid_radius;
+    if (x_max >= 1.0F and y_max >= 1.0F) grid_radius = std::max(x_max, y_max);
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: we do not use and we use &&

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 3]
```diff
@@ -236,7 +251,8 @@ bool DeltaGridStrategy::probe_grid(int n, float radius, StreamOutput *stream)
             // Avoid probing the corners (outside the round or hexagon print surface) on a delta printer.
             float distance_from_center = sqrtf(x*x + y*y);
             float z= 0.0F;
-            if (distance_from_center <= radius) {
+            if ((!is_square and (distance_from_center <= radius)) or
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: we don't use or we use ||

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 4]
```diff
@@ -236,7 +251,8 @@ bool DeltaGridStrategy::probe_grid(int n, float radius, StreamOutput *stream)
             // Avoid probing the corners (outside the round or hexagon print surface) on a delta printer.
             float distance_from_center = sqrtf(x*x + y*y);
             float z= 0.0F;
-            if (distance_from_center <= radius) {
+            if ((!is_square and (distance_from_center <= radius)) or 
+               ((x < -x_max or x > x_max or y < -y_max or y > y_max) and is_square)) {
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: is_square should be first to short circuit the following test

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 5]
```diff
@@ -453,9 +467,12 @@ bool DeltaGridStrategy::doProbe(Gcode *gc)
         for (int xCount = xStart; xCount != xStop; xCount += xInc) {
             float xProbe = LEFT_PROBE_BED_POSITION + AUTO_BED_LEVELING_GRID_X * xCount;
 
+            // avoid probing outside of x min/max on a cartesian
+            if ((xProbe < -x_max or xProbe > x_max or yProbe < -y_max or yProbe > y_max) and is_square) continue;
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: is_square first to short circuit other test

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 6]
```diff
@@ -109,6 +113,17 @@ bool DeltaGridStrategy::handleConfig()
     grid_size = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, grid_size_checksum)->by_default(7)->as_number();
     tolerance = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, tolerance_checksum)->by_default(0.03F)->as_number();
     save = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, save_checksum)->by_default(false)->as_bool();
+    x_max = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, x_max_checksum)->by_default(0.0F)->as_number();
+    y_max = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, y_max_checksum)->by_default(0.0F)->as_number();
+    do_home = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, do_home_checksum)->by_default(true)->as_bool();
+    is_square = THEKERNEL->config->value(leveling_strategy_checksum, delta_grid_leveling_strategy_checksum, is_square_checksum)->by_default(false)->as_bool();
+
+    // intelligently set defaults.
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: Should probably wrap this in is_square as no need to do it if it is a delta

File: src/modules/tools/zprobe/DeltaGridStrategy.cpp [hunk 7]
```diff
@@ -236,7 +251,8 @@ bool DeltaGridStrategy::probe_grid(int n, float radius, StreamOutput *stream)
             // Avoid probing the corners (outside the round or hexagon print surface) on a delta printer.
             float distance_from_center = sqrtf(x*x + y*y);
             float z= 0.0F;
-            if (distance_from_center <= radius) {
+            if ((!is_square && (distance_from_center <= radius)) ||
+               ((x < -x_max || x > x_max || y < -y_max || y > y_max) && is_square)) {
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/DeltaGridStrategy.cpp]: still need to put is_square first, as it short circuits the big or test

**Comment by dethegeek** [comment]
Hi

I did some tests on my (cartesian) printer. Probing behaves as expected on a square bed. Would be useful to go further and enable rectangular bed (mine has Y slightly larger than X).

**Comment by lordofhyphens** [comment]
Yeah it should behave the same on a rectangular bed, although you may want
to increase the number of probes. The rectangular setup basically just
doesn't probe part of a larger square.
On May 29, 2016 9:02 AM, "dethegeek" <notifications@github.com> wrote:

Hi

I did some tests on my (cartesian) printer. Probing behaves as expected on
a square bed. Would be useful to go further and enable rectangular bed
(mine has Y slightly larger than X).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Smoothieware/Smoothieware/pull/907#issuecomment-222362225>,
or mute the thread
<https://github.com/notifications/unsubscribe/AAB8CvG68VKhMa9GvbQTVibYQj1RWeKIks5qGZx1gaJpZM4IEBUW>
.

**Comment by pflugj** [comment]
Hi and first of all:
Thanks for all your efforts and hard work !!

I am also trying to test/implement the delta-grid leveling strategy with my "Voxel OX" cartesian printer.
...but I am doing something wrong and after days of searching I can not figure out what is wrong ;-(

I am running the latest smoothie edge from the 1st of June
I had 3-point leveling enabled and everything worked fine.

So I enabled delta-grid like so:
##################  Leveling Strategy "delta-grid"
leveling-strategy.delta-grid.enable                   true
leveling-strategy.delta-grid.radius                   100
leveling-strategy.delta-grid.size                     7
leveling-strategy.delta-grid.probe_offsets            60,40,0  # probe offsets x,y,z
leveling-strategy.delta-grid.save                     true
leveling-strategy.delta-grid.initial_height           5
leveling-strategy.delta-grid.wait_for_probe           false         # Makes system wait for probe confirmation

and added 
#########  new config options to leveling-strategy.delta-grid:
max_x                                            100     # defaults to value of grid_radius - Maximum travel from center of bed in +/-X direction
max_y                                            100     # defaults to value of grid_radius - Maximum travel from center of bed in +/-Y direction.
do_home                                          false   # default True - Home at beginning of G31 call.
is_square                                        true    # default False - turns off the check to not probe the corners (do not set on a delta with a round bed)
#CAUTION:0.0 must be center of the bed when using this leveling strategy.

To make sure that the bed center is at "0.0"  before leveling
I issue the command to offset the P1 plane 
G10 L2 P1 X144 Y 154 Z0
after G28 homing. 
This puts the nozzle right at the center of the print bed @ 0,0 and 
at the "machines home" position the LCD displays X-144 Y-154 Z0.00

Now when I issue a G29 command I assumed it would the probe a grid of 7x7 points within the +-100mm X and Y limits from the bed center
But what happens is:
The printer homes to the "machine home" ( X-144 Y-154) and then lowers Z slamming the nozzle right into the bed!

So what am I doing wrong here?
What would be the correct procedure to use this strategy?

Again, thanks for helping

Thanks for all your hard work and help
Jurgen

**Comment by lordofhyphens** [comment]
This calibration routine is for G31, not G29. See if that helps.

If your printer can home normally and then go to 0,0 in the center of the
bed (either through use of M206 or setting alpha_min/max and beta_min/max)
then you can set do_home = True.

"Nothing unreal exists." - Kiri-kin-tha's First Law of Metaphysics.

**Comment by pflugj** [comment]
upps ...that was a quick response ;-)

I have changed alpha and beta min/max
and changed do_home to "true".
After G28 and G0 X0 Y0 the nozzle is now also in the center ;-)

Now I am using "G31" and the terminal shows:
"Delta Grid Probe...
still the printer returns to the "machines home" and slams right into the bed ;-( 

Any other ideas?

**Comment by wolfmanjm** [comment]
you cannot change bed center to 0,0 by using G10. the bed has to be center in machine coordinates NOT WCS coordinates. You would adjust the print space to whatever you want afterwards with G10 though. but the machine coordinates center must be 0,0

Also did you apply the PR and rebuild? the default edge does not support cartesians for delta grid.

**Comment by pflugj** [comment]
Upps, I now see where the error is:
I am just using the '"default edge" 

I do not know how to apply the "PR" (is that a..."P"ull "R"equest?)
.. and I certainly did not rebuild ;-(
Is there anywhere a tutorial how to do this things?
...or will that strategy be implemented anytime soon?

Again thanks for you help

**Comment by lordofhyphens** [comment]
https://github.com/lordofhyphens/Smoothieware/tree/deltagrid_cartesian if you want a download of that. You will have to build it from source (see the Smoothieware wiki IIRC for those instructions).

Alternatively, you can fork your own copy of Smoothieware, and start a pull request on that from lordofhyphens:deltagrid_cartesian (Github has tutorials for this), then download *that* copy and build the firmware. This has the advantage of including everything that was in edge after I made the PR to make sure nothing else broke.

**Comment by pflugj** [comment]
...looks like I will have another sleepless night then ;-)

Thanks for you help

**Comment by lordofhyphens** [comment]
It only really takes 10-15m to build if you're following the instructions.

**Comment by pflugj** [comment]
maybe only 10-15 min after everything is on my laptop.
....current download speed almost 0kb/sec
....this is Africa!!

**Pull request merged**